overlord/state/taskrunner.go

```diff
@@ -184,6 +186,19 @@ func (r *TaskRunner) Ensure() {
 	}
 }
 
+// wait expectes to be called with th r.mu lock held
+func (r *TaskRunner) wait() {
+	for len(r.tombs) > 0 {
+		for id, t := range r.tombs {
+			r.mu.Unlock()
+			t.Wait()
+			r.mu.Lock()
+			delete(r.tombs, id)
```

Comment:
By the time the t.Wait above returns, the task must have returned its function, and the task id shouldn't be there anymore per the changes above.